cloudprovider/providers/aws/aws.go

```diff
@@ -199,21 +203,54 @@ func NewProvider(awsServices Services, az string) (cloudTypes.Interface, error)
 	return &provider{availabilityZone: az, region: region, ec2: ec2, metadata: metadata}, nil
 }
 
-func newAWSSDKProvider(creds *credentials.Credentials, maxRetries int) *awsSDKProvider {
-	return &awsSDKProvider{creds: creds, maxRetries: maxRetries}
+func newAWSSDKProvider(config *aws.Config, ec2Metadata *ec2metadata.EC2Metadata) (Services, error) {
+	if config.HTTPClient != nil && config.HTTPClient.Timeout <= 0 {
+		return nil, errors.New("http client timeout must be positive")
+	}
+	return &awsSDKProvider{
+		config:      config,
+		ec2Metadata: ec2Metadata,
+	}, nil
 }
 
 // NewProviderFromViper returns a new aws provider.
 func NewProviderFromViper(v *viper.Viper) (cloudTypes.Interface, error) {
-	v.SetDefault("aws.max_retries", 3)
-	creds := credentials.NewChainCredentials(
-		[]credentials.Provider{
-			&credentials.EnvProvider{},
-			&ec2rolecreds.EC2RoleProvider{
-				Client: ec2metadata.New(session.New(&aws.Config{})),
-			},
-			&credentials.SharedCredentialsProvider{},
-		})
-	aws := newAWSSDKProvider(creds, v.GetInt("aws.max_retries"))
-	return NewProvider(aws, v.GetString("aws.availability_zone"))
+	a := getSubViper(v, "aws")
+	a.SetDefault("max_retries", 3)
+	a.SetDefault("http_timeout", 3*time.Second)
+	// This is the main config without credentials.
+	config := &aws.Config{
+		MaxRetries: aws.Int(a.GetInt("max_retries")),
+		HTTPClient: &http.Client{
+			Timeout: a.GetDuration("http_timeout"),
+		},
```

Comment:
This http client is going to be used for all requests. Currently we do not specify timeout at all and this is bad. Metadata service sets it to 5 seconds by default, but other services may just hang for a long time blocking Receiver goroutine unnecessary.